### PR TITLE
Name parsing fixes

### PR DESCRIPTION
**Pull request by jaydenseric**
Fixes https://github.com/syavorsky/comment-parser/issues/83 and fixes https://github.com/syavorsky/comment-parser/issues/84 and fixes https://github.com/syavorsky/comment-parser/issues/86 .

This is a more comprehensive fix than https://github.com/syavorsky/comment-parser/pull/85 .

### CODE REVIEW

File: parsers.js [hunk 1]
```diff
@@ -74,13 +74,14 @@ PARSERS.parse_name = function parse_name (str, data) {
...
-      name = name.slice(1, -1)
 
-      if (name.indexOf('=') !== -1) {
-        const parts = name.split('=')
-        name = parts[0]
-        res.default = parts[1].replace(/^(["'])(.+)(\1)$/, '$2')
-      }
+      const result = name.slice(1, -1).match(
+        /^\s*([^=]+?)(?:\s*=\s*(.+?))?\s*(?=$)/
+      )
```
Comment (syavorsky) [inline parsers.js:80]: @jaydenseric good point, this should really be less restrictive. Would you mind if we avoid the RegExp to keep it simpler. Maybe something like
```js
name = name.slice(1, -1)
const eq = indexOf('=')
if (eq !== -1) {
  name = name.slice(0, eq).trim()
  res.default = name.slice(eq + 1).trim()
}
```
other than that I would merge it right away

Comment (jaydenseric) [reply parsers.js:80]: I can sympathise with regex being a little tricky to follow, but one of the nice things about having a comprehensive regex is that you can tell up front if the name is valid syntax or not. I hadn't done it, but you could do something like:

```js
const result = name.slice(1, -1).match(/^\s*([^=]+?)(?:\s*=\s*(.+?))?\s*(?=$)/);

// Add this:
if (!result) throw new SyntaxError('Invalid syntax for a name in a JSDoc tag.');
```

For example, what if the name is `[=]`? Or `[==]`? IMO it's better to throw a syntax error than return a misleading `''` or `=` name. Without a regex you would have to do a lot of little checks throught the code.

Comment (jaydenseric) [reply parsers.js:80]: I just added a few more tests to the regex proving the invalid syntax names don't match: https://regex101.com/r/owjcTx/7/tests .

**Comment by jaydenseric** [comment]
Honestly I think I'll just make my own JSDoc parsing lib.

File: parsers.js (hunk 1 above)
Comment (syavorsky) [reply parsers.js:81]: Makes sense, let’s keep the regexp

File: tests/parse.spec.js [hunk 2]
```diff
@@ -458,52 +460,117 @@ describe('Comment string parsing', function () {
...
-          tag: 'my-tag',
-          line: 2,
-          type: 'my.type',
-          name: 'name',
-          description: '',
-          source: '@my-tag {my.type} [name]',
-          optional: true
-        }]
-      })
+        optional: false,
```
Comment (jaydenseric) [inline tests/parse.spec.js:482]: I'm not really happy that it says that optional is `false`, when really it was true. The problem is there is no mechanism to throw a syntax error at that point in the parser without losing the optional data.

File: tests/parse.spec.js [hunk 3]
```diff
@@ -458,52 +460,117 @@ describe('Comment string parsing', function () {
...
-      .to.eql({
+    })).to.eql([{
+      line: 1,
+      description: '',
+      source: '@my-tag [=value]',
+      tags: [{
+        tag: 'my-tag',
+        line: 2,
+        type: '',
+        name: '',
```
Comment (jaydenseric) [inline tests/parse.spec.js:501]: It also doesn't sit well that `type` and `name` have `''` values when they were really undefined, but these are design decisions beyond the scope of this PR.

**Pull request merged**